### ISSUE TITLE
Improve Node Grouping in Kedro Deployment

### ISSUE THREAD

**Issue by DimedS** [kedro-org/kedro#4319]
## Overview
Part of #4317. Users have expressed the need to merge multiple Kedro nodes into a single task on deployment platforms for better clarity and efficiency. Current plugins offer limited support for this, often requiring manual grouping, which complicates deployment and reduces performance.

## User Insights and Challenges
- "Combining nodes into single tasks improves overview, but we currently have to manually group them in Databricks."
- "We can convert a single node to a Kubeflow Component, but deploying 400 nodes as separate containers adds complexity."
- "Running each Kedro node in a separate container could make a small node execute in one or two seconds, but Argo’s longer pod startup time would make this inefficient."

## Problem Statement
How can we design a flexible and efficient node grouping mechanism - using tags, namespaces, pipelines, or other methods - to maximise usefulness for users and streamline the deployment process?

## Proposed Solution
- **Centralised Grouping Functionality**: Instead of developing node grouping features separately for each plugin, centralise this functionality within the Kedro framework. This approach would standardise and simplify node grouping, making it easier to implement and maintain across different deployment platforms.

**Comment by datajoely** [comment]
Users today also tend towards tags because namespaces are a pain to use

**Comment by DimedS** [comment]
## Summary of Tech Design Sessions on Node Grouping (4, 6 Dec 2024)

### Discussed:
- **Node Grouping Methods:** Reviewed different methods for grouping nodes in Kedro, including tags, pipelines, and namespaces. Discussed the pros and cons of each approach and how they affect deployment when converting Kedro pipelines to various platforms.
- **Deployment Options:** Considered two options for deploying grouped nodes:
  1. Helper functions within Kedro combined with modifications to deployment plugins.
  2. Using synthetic nodes.

### Decisions:
1. **Adopting Namespaces for Deployment:**
   - **Reasoning:**  
     Despite observed challenges with namespaces (e.g., limited adoption and UI complexities), namespaces are considered the right tool for this use case. They offer value to users and should be more widely utilized. Enhancing deployment functionality based on namespaces, alongside improving the UI, will likely boost their adoption.
2. **Choosing Helper Functions + Plugins Over Synthetic Nodes:**
   - **Reasoning:**  
     Synthetic nodes are technically complex, less useful for users, and harder to implement. Instead, focusing on helper functions and plugin modifications aligns better with Kedro’s approach and encourages collaboration with plugin maintainers to ensure proper maintenance.

### Next Steps:
1. **Prototype Development:**
   - Develop a deployment solution prototype using helper functions in Kedro, tested with the `kedro-airflow` plugin to enable namespace deployment to Airflow. Results will be shared in the Tech Design session, approximately in January.
2. **Namespace UI Improvements:**
   - Collaborate with Kedro-Viz (@Huongg) to improve the UI for namespaces and drive increased adoption. (See issue #4343.)
3. **Engagement with Plugin Developers:**
   - Continue engaging with deployment plugin developers to establish a stable plugin maintenance model. (See issue #4318.)

### Additional Information:
Feel free to add your thoughts or suggestions here. If there’s anything to update or clarify in the summary, please let me know.  
A link to the video recordings of both sessions will be published soon.

**Comment by DimedS** [comment]
from @marrrcin :
_I was not able to attend the follow up :sadcat2:
For plugin development, a functionality that would be super helpful would be an utility function in official Kedro API that will allow to pass in the:
pipeline
grouping criterion
and result in list of grouped nodes on the grouping criterion.
I don't know if that's how you plan to expose that :thinking_face:
By iterating on the results of such grouped pipeline, plugins will be able to easily do the M:N mapping instead of 1:1 . Then plugins will be able to create orchestrator-specific pipelines which will invoke sth like k`edro run --from-nodes --to-nodes` ._

**Comment by astrojuanlu** [comment]
> Despite observed challenges with namespaces (e.g., limited adoption and UI complexities), namespaces are considered the right tool for this use case.

I value the desire of keeping momentum and devoting more time to understanding what namespaces do and how do they work so at least we can discuss more intelligently about them.

But going forward I think there's an opportunity to explore innovative new solutions, or just make namespaces an implementation detail so that they continue to exist but they become invisible for the user and get swept under a more usable API layer. I know I sound like a broken record but I'll say it again: more documentation will not fix bad Developer Experience.

In short: agree to continue exploring them (on the grounds of keeping the momentum on this topic and not having to do another knowledge sharing session in ~12 months), but I think we should timebox this effort, and put a deadline on when do we think we're ready to go back to the drawing board and continue iterating as a team.

**Comment by datajoely** [comment]
@marrrcin @Galileo-Galilei @sbrugman  would be interested if you also agree that dependency isolation is something we need to also consider within this problem space?

**Comment by marrrcin** [comment]
What do you mean by that exactly?

**Comment by datajoely** [comment]
I think namespaces are critical for any way we eventually unify and simplify deployment to orchestrators, I'm arguing dependency isolation is the basically the same problem - but would like to see if others agree?

**Comment by deepyaman** [comment]
> @marrrcin @Galileo-Galilei @sbrugman would be interested if you also agree that dependency isolation is something we need to also consider within this problem space?

I think it's a valid consideration. At the very least, there needs to be an answer as to how you handle pipelines that have multiple sets of dependencies (e.g. should `data_engineering` and `data_science` be separate pipelines, or should they be separate Kedro projects, if they have a separate set of, potentially-incompatible, dependencies?). In my past work on large Kedro projects (ages ago), I heavily leveraged monorepo approaches, as in #4147, because there is also a tradeoff to maintaining multiple Kedro projects for a single use case.

I can't find my previous comments to this extent (perhaps they were verbal), but my opinions are:

* Micro-packaging was always a bad idea (for Kedro framework to support); they were introduced to solve a QB-internal use case that never even leveraged them/built their own solution.
* However, being able to define separate requirements for top-level pipelines IS useful, especially for deployment. The story for running a pipeline locally with multiple sets of dependencies was never really there, but it seems useful.
* Supporting complex node grouping is unnecessary, at least to start with; that's akin to the micro-packaging over-engineering problem. The vast majority of users would benefit from being able to deploy each modular pipeline (with it's own namespace) separately. This also creates pretty clear boundaries where node persistence, etc. may be required.

TL;DR I agree with the idea of namespaces and supporting deployment based on namespaces, but I further posit that namespaces just being per modular pipeline is sufficient, even for deployment, and that will also make them easier to adopt.

**Comment by Galileo-Galilei** [comment]
I mostly agree with all of deepyaman's comments. I actually crave for pipelines requirements isolation, but I've never thought about requirements isolation at namespace level. This would make sense from an orchestrator point of view with "big nodes" filtered out from namespaces pipelines, but I am not sure this is what we should focus on in the short term because it's likely a hard problem. 

IMO the short term priority is to increase developer experience and increase usage/ discoverability. I can't agree more with what @astrojuanlu said about no documentation can catch up a not ideal developer experience.

**Comment by marrrcin** [comment]
I agree with the need for dependency isolation for large projects 👍🏻

For grouping by the namespace - although it's fine to have anything to group on, namespaces were (at least for me) used to group larger chunks of the pipeline, e.g.:
```
<ns1 = data processing part> --> <ns2 = modeling variant 1>
                                                   --> <ns2 = modeling variant 2>
```
etc. (like in the dynamic pipelines https://getindata.com/blog/kedro-dynamic-pipelines/ ).

So grouping by the namespace will actually require thinking about the target deployment as soon as you start writing the pipeline, which means it will impact the data catalog/parameters creation too (because of namespace prefixes) = more cognitive load.

Plus the projects might end up with having a lot of "synthetic" namespaces, just for the sake of preparing the pipeline for an orchestrator.

**Comment by astrojuanlu** [comment]
Copy-pasting some comments about dependency isolation to https://github.com/kedro-org/kedro/discussions/4147 and collapsing them here

**Comment by deepyaman** [comment]
> Copy-pasting some comments about dependency isolation to #4147 and collapsing them here

OK, but I'm going to just reiterate my points specifically related to node grouping and deployment, so that they don't get skipped/misconstrued as being only relevant for dependency management:

> * Supporting complex node grouping is unnecessary, at least to start with; that's akin to the micro-packaging over-engineering problem. The vast majority of users would benefit from being able to deploy each modular pipeline (with it's own namespace) separately. This also creates pretty clear boundaries where node persistence, etc. may be required.
> 
> TL;DR I agree with the idea of namespaces and supporting deployment based on namespaces, but I further posit that namespaces just being per modular pipeline is sufficient, even for deployment, and that will also make them easier to adopt.

**Comment by astrojuanlu** [comment]
Next steps:

1. Write a guide that covers all the existing node grouping strategies for deployment, including namespaces
2. Implement communications strategy to radiate the message (a blog post, a video)
3. Perform some user testing to get user feedback from existing strategies
4. Synthesise findings and decide way forward

**Comment by Huongg** [comment]
Thanks, @astrojuanlu, for the summary—it sounds great! We've also drafted a problem statement, which we hope will help us focus more on the goal. What do you think? 

**"How might we support users in optimizing their code for deployment by effectively grouping their nodes and providing clear, actionable guidance through documentation?"**

**Comment by iamelijahko** [comment]
Here are the 3 users response on **Node Grouping Documentation Validation** that Kedro team conducted in April 2025:
**Documentation page**: https://docs.kedro.org/en/stable/deployment/nodes_grouping.html

###  What We Are Validating?
| Validation Point | Actions | User 1 | User 2 | User 3 |
| --- | --- | --- | --- | --- |
| Information architecture | Issue 1: Reconsider Placement of "Nodes Grouping" Documentation in the Kedro Docs | N/A | N/A | Is **Advanced usage → Deployment** the best place to add this? To me, this page looks like it's listed among a bunch of different stacks where we can deploy Kedro... <br></br> The reason I ask is because I see this more as a **design decision made during the initial development phase** of a Kedro project, rather than something related to actual deployment (which comes at the end). Personally, I would never think to look for it under "Deployment" — I’d expect it under **Configuration** or something similar. <br></br> Here’s how I think about it: <br></br> When I’m setting up a Kedro project, the decision to split nodes into pipelines, tags, or namespaces happens early on, as part of how I structure the code. <br></br> • I create Pipeline() objects for pipeline grouping <br></br> • I add tags=[] to nodes for tag-based grouping <br></br> • I define namespaced pipelines if I want to use namespaces <br></br> All of these are **code-level**, **structural decisions** made during development/configuration time. <br></br> **Only after** the project is complete do I start thinking about **where to deploy it**, and that’s when I’d naturally turn to the Deployment docs — e.g., for Databricks, Prefect, etc. |
| Grouping by pipelines | Issue 2: Clarify Connection Between Node Grouping Methods and `register_pipelines()` in the Docs | N/A | • I like the section. Very straightforward. <br></br> • Maybe adding that specifically those are the ones you add into the pipeline registry (?) if that is not too much detail <br></br> Kedro: Great callout — yes, those are indeed the ones you register in the pipeline registry (`register_pipelines()`). We’ll clarify that explicitly in the docs to connect the dots better! | N/A |
| Grouping by tags | **Issue 3: Clarify Limitations of Using Tags with Node Dependencies in Kedro Docs** <br></br> The current Nodes Grouping guide includes a note that tags should not be used when "tagged nodes have strong dependencies, which might cause execution failures." However, this explanation is vague and could be made clearer with concrete examples. For example, if a tagged node depends on a `MemoryDataSet`, and the node producing that dataset is not tagged and therefore not run, Kedro will fail at runtime. <br></br> **Issue 4: Improve Explanation of Tag Limitations in the Node Grouping Guide** <br></br> **i. Clarify what “not hierarchical” means:** Update the guide to explicitly state:Tags are not hierarchical like pipelines or namespaces — they are flat labels. This makes it harder to organize and trace complex workflows, especially in large or collaborative projects. <br></br> **ii. Explain the implications for teams:** Add a note like:In team settings, inconsistent or arbitrary tag usage can make it hard to understand how nodes relate to one another. Unlike pipelines or namespaces, tags don’t enforce structure or clarity. <br></br> **iii. Contrast with other grouping methods:** Create a short comparison block showing how tags differ from pipelines/namespaces: <br></br> **Issue 5: Clarify Use Cases and Benefits of Tags in the Node Grouping Guide** <br></br> Recent user feedback suggests that tags — while flexible — can feel risky and of low utility compared to more structured grouping methods like pipelines or namespaces. The current documentation emphasizes their limitations, which may discourage their use without offering clear, compelling examples. | ```Not to use when • The tagged nodes have strong dependencies, which might cause execution failures.``` <br></br> **User 2**: I think this is saying "The tagged nodes have MemoryDatasets as inputs. These inputs will not be created when running only the tagged nodes." <br></br> **Kedro**: You're right—that's one example of what could happen. <br></br> ```Not to use when • Tags are not hierarchical, so tracking groups of nodes can become difficult.``` <br></br> **User 2**: Is this saying "Tags do not necessarily compose a DAG"? <br></br> **Kedro**: I think the key point here is not about the DAG itself, but rather how tags are used to group nodes. Two main challenges we’ve identified are: <br></br>• Tags are not inherently hierarchical like pipelines or namespaces. If multiple developers assign tags arbitrarily, it can be difficult to track which nodes belong together. <br></br> • Tags do not enforce structure—they are just labels. A new developer joining the project might not immediately understand which nodes are critical for different deployments. <br></br> If we explain a bit more about these two points as above, would it make more sense in the guide to you? <br></br> **User 2**: I think that would help. I've always worked solo on Kedro projects, so the "If multiple developers..." part was eye-opening. The second point is clear, too. | • Tags are tricky, and from the description I'd say that it does not encourage too much to use them. Like low reward and high risk <br></br> • Seems like a niche problem 🤔. Like I would use them if I have nodes in separate pipeline, but could that be a hint that my pipeline is not structured correctly? I would love to hear creative ways on how other people have use them, other than metadata <br></br> **Kedro**: Totally fair observation. Tags can feel a bit risky when used for execution because they’re not structured or validated like pipelines/namespaces. They’re more useful for **visual organization, debugging, or temporary grouping** across pipelines — especially when you're not ready to restructure yet. We’ll see if we can surface more examples or creative use cases to help make their value clearer! | N/A |
| Grouping by namespaces | **Issue #6: Improve Documentation and Validation for Namespaces in Node Grouping** <br></br> • Clarify the intended usage of namespaces in the docs by recommending pipeline-level namespaces over node-level ones <br></br> • Providing practical examples of when to use kedro run --namespace=... <br></br> • Warning about execution limitations (e.g., with MemoryDataSet inputs), improving validation tooling to detect misuses | N/A | • Grouping by pipelines is inevitable, but using namespaces is the most important for me otherwise haha they feel more impactful when trying to organize the pipeline <br></br> • Namespaces at node-level seems more like a weird behavior, as an unintended side-effect from the implementation. Is there a reason it exists? Should it be a hidden property? That I don't know <br></br> • When I try to explain other people how to use namespaces, I like to tell them that is like adding a last name. <br></br> • Now that I see `kedro run --namespace=<your_namespace_name>`, what scenario is there where I would prefer running a namespace over a pipeline? I would also need to clearly define the I/O of my namespace, because if the inputs are MemoryDataset, it would fail the run <br></br> **Kedro**: <br></br> • Completely agree — namespaces often end up being the most powerful way to organize complex pipelines. <br></br> • Your “last name” metaphor is great — love that for explaining it to others! <br></br> • You're also spot on that node-level namespaces are more of an implementation artifact. They’re technically allowed, but the safer and recommended path is to apply them at the pipeline level, where execution can be validated (e.g., using `kedro run --namespace=...`). We’re already working on validation improvements to help catch misuses early on. <br></br> • As for your question on when you'd run a namespace instead of a pipeline — it's mainly when you want to target a specific subset of a larger pipeline. For example, in deployment environments where each namespace maps to a task or container, running a namespace can reduce overhead compared to running the full pipeline. But you’re right — it only works well when the namespace’s inputs/outputs are properly defined (i.e., not just in-memory). | N/A |
